pyomo/contrib/interior_point/interior_point.py

```diff
@@ -0,0 +1,669 @@
+    alpha_dual_max: float
+    """
+    primals = interface.get_primals()
+    slacks = interface.get_slacks()
+    duals_primals_lb = interface.get_duals_primals_lb()
+    duals_primals_ub = interface.get_duals_primals_ub()
+    duals_slacks_lb = interface.get_duals_slacks_lb()
+    duals_slacks_ub = interface.get_duals_slacks_ub()
+
+    delta_primals = interface.get_delta_primals()
+    delta_slacks = interface.get_delta_slacks()
+    delta_duals_primals_lb = interface.get_delta_duals_primals_lb()
+    delta_duals_primals_ub = interface.get_delta_duals_primals_ub()
+    delta_duals_slacks_lb = interface.get_delta_duals_slacks_lb()
+    delta_duals_slacks_ub = interface.get_delta_duals_slacks_ub()
+
+    primals_lb = interface.primals_lb()
+    primals_ub = interface.primals_ub()
+    ineq_lb = interface.ineq_lb()
+    ineq_ub = interface.ineq_ub()
+
+    alpha_primal_max_a = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=primals,
+        delta_x=delta_primals,
+        xl=primals_lb)
+    alpha_primal_max_b = _fraction_to_the_boundary_helper_ub(
+        tau=tau,
+        x=primals,
+        delta_x=delta_primals,
+        xu=primals_ub)
+    alpha_primal_max_c = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=slacks,
+        delta_x=delta_slacks,
+        xl=ineq_lb)
+    alpha_primal_max_d = _fraction_to_the_boundary_helper_ub(
+        tau=tau,
+        x=slacks,
+        delta_x=delta_slacks,
+        xu=ineq_ub)
+    alpha_primal_max = min(alpha_primal_max_a, alpha_primal_max_b,
+                           alpha_primal_max_c, alpha_primal_max_d)
+
+    alpha_dual_max_a = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=duals_primals_lb,
+        delta_x=delta_duals_primals_lb,
+        xl=np.zeros(len(duals_primals_lb)))
+    alpha_dual_max_b = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=duals_primals_ub,
+        delta_x=delta_duals_primals_ub,
+        xl=np.zeros(len(duals_primals_ub)))
+    alpha_dual_max_c = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=duals_slacks_lb,
+        delta_x=delta_duals_slacks_lb,
+        xl=np.zeros(len(duals_slacks_lb)))
+    alpha_dual_max_d = _fraction_to_the_boundary_helper_lb(
+        tau=tau,
+        x=duals_slacks_ub,
+        delta_x=delta_duals_slacks_ub,
+        xl=np.zeros(len(duals_slacks_ub)))
+    alpha_dual_max = min(alpha_dual_max_a, alpha_dual_max_b,
+                         alpha_dual_max_c, alpha_dual_max_d)
+
+    return alpha_primal_max, alpha_dual_max
+
+
+def process_init(x, lb, ub):
+    if np.any((ub - lb) < 0):
+        raise ValueError(
+            'Lower bounds for variables/inequalities should not be larger than upper bounds.')
+    if np.any((ub - lb) == 0):
+        raise ValueError(
+            'Variables and inequalities should not have equal lower and upper bounds.')
+
+    lb_mask = build_bounds_mask(lb)
+    ub_mask = build_bounds_mask(ub)
+
```

Comment:
Do these masks not exist already somewhere?

Reply:
Good question. I originally had a bunch of compression matrices floating around with lots of matrix vector dot products. However, when I started profiling, those matrix vector dot products were the first thing that showed up. It turns out that, throughout most of the algorithm, numpy handles the bounds of `-inf` and `inf` quite nicely. In most parts of the algorithm, we are dividing by `x - x_L` or `x_U - x`. When the denominator is `inf`, numpy returns `0`, so things work out nicely. Because of this, I just used the full arrays of primals, duals, etc. throughout the algorithm instead of compressed arrays. The duals for variables with bounds are just always `0`. This just leaves a couple places where we have to worry about infinite bounds: when processing initial values and in the convergence check.